util/src/logging/formatter.rs

```diff
@@ -0,0 +1,174 @@
+//! to the `dd.span_id` field, which is where Datadog looks for these by default
+//! (although the path to the trace ID can be overridden in Datadog).
+
+#![allow(missing_docs)]
+#![allow(clippy::missing_docs_in_private_items)]
+
+use std::io;
+
+use chrono::Utc;
+use opentelemetry::trace::{SpanId, TraceContextExt, TraceId};
+use serde::ser::{SerializeMap, Serializer as _};
+use serde::Serialize;
+use tracing::{Event, Subscriber};
+use tracing_opentelemetry::OtelData;
+
+use tracing_serde::AsSerde;
+use tracing_subscriber::fmt::format::Writer;
+use tracing_subscriber::fmt::{FmtContext, FormatEvent, FormatFields};
+use tracing_subscriber::registry::{LookupSpan, SpanRef};
+
+#[derive(Serialize)]
+struct DatadogId(u64);
+
+struct TraceInfo {
+    trace_id: DatadogId,
+    span_id: DatadogId,
+}
+
+impl From<TraceId> for DatadogId {
+    fn from(value: TraceId) -> Self {
+        let bytes = &value.to_bytes()[std::mem::size_of::<u64>()..std::mem::size_of::<u128>()];
+        Self(u64::from_be_bytes(bytes.try_into().unwrap_or_default()))
+    }
+}
+
+impl From<SpanId> for DatadogId {
+    fn from(value: SpanId) -> Self {
+        Self(u64::from_be_bytes(value.to_bytes()))
+    }
+}
+
+fn lookup_trace_info<S>(span_ref: &SpanRef<S>) -> Option<TraceInfo>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    span_ref.extensions().get::<OtelData>().map(|o| {
+        let trace_id = if o.parent_cx.has_active_span() {
+            o.parent_cx.span().span_context().trace_id()
+        } else {
+            o.builder.trace_id.unwrap_or(TraceId::INVALID)
+        };
+        TraceInfo {
+            trace_id: trace_id.into(),
+            span_id: o.builder.span_id.unwrap_or(SpanId::INVALID).into(),
+        }
+    })
+}
+
+// mostly stolen from here: https://github.com/tokio-rs/tracing/issues/1531
+pub struct DatadogFormatter;
+
+impl<S, N> FormatEvent<S, N> for DatadogFormatter
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_event(
+        &self,
+        ctx: &FmtContext<'_, S, N>,
+        mut writer: Writer<'_>,
+        event: &Event<'_>,
+    ) -> std::fmt::Result
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        let meta = event.metadata();
+
+        let mut visit = || {
+            let mut serializer = serde_json::Serializer::new(WriteAdaptor::new(&mut writer));
+            let mut serializer = serializer.serialize_map(None)?;
+            serializer.serialize_entry("timestamp", &Utc::now().to_rfc3339())?;
+            serializer.serialize_entry("level", &meta.level().as_serde())?;
+            serializer.serialize_entry("target", meta.target())?;
+
+            if let Some(filename) = meta.file() {
+                serializer.serialize_entry("filename", filename)?;
+            }
+
+            if let Some(line_number) = meta.line() {
+                serializer.serialize_entry("line_number", &line_number)?;
+            }
```

Comment:
This file was copied form https://github.com/will-bank/datadog-tracing/blob/main/src/formatter.rs except for these lines, which were added (and themselves taken from https://github.com/tokio-rs/tracing/blob/master/tracing-subscriber/src/fmt/format/json.rs#L260-L270)